src/cpu_treeshap_model.h

```diff
@@ -0,0 +1,420 @@
+            rapids::Error::Unsupported, "Model does not have node statistics.");
+      }
+      if (seen_features.count(feature)) {
+        auto seen_node_id = seen_features[feature];
+        parents[node] = seen_node_id;
+        weight *= weights[seen_node_id];
+      }
+      weights[node] = weight;
+      seen_features[feature] = node;
+    }
+
+    double bias = 0.0;
+    if (!tree.IsLeaf(node)) {
+      auto [left_max_features, left_bias] =
+          Recurse(tree.LeftChild(node), node, depth + 1, seen_features);
+      auto [right_max_features, right_bias] =
+          Recurse(tree.RightChild(node), node, depth + 1, seen_features);
+      edge_heights[node] = std::max(left_max_features, right_max_features);
+      bias = left_bias + right_bias;
+    } else {
+      edge_heights[node] = seen_features.size();
+      max_depth = std::max(max_depth, depth);
+      double leaf_value = tree.HasLeafVector(node)
+                              ? tree.LeafVector(node)[class_idx]
+                              : tree.LeafValue(node);
+      bias = leaf_probability(tree, node) * leaf_value;
+    }
+    return std::make_pair(edge_heights[node], bias);
+  }
+
+  explicit TreeMetaInfo(
+      const treelite::Tree<tl_threshold_t, tl_output_t>& tree, int class_idx)
+      : tree(tree), class_idx(class_idx)
+  {
+    weights.resize(tree.num_nodes, 1.0f);
+    parents.resize(tree.num_nodes, -1);
+    edge_heights.resize(tree.num_nodes);
+    auto [_, bias] = Recurse(0);
+    global_bias = bias;
+    if (max_depth >= 32) {
+      throw rapids::TritonException(
+          rapids::Error::Unsupported,
+          "Tree depths above 32 not supported. Shap cannot be considered "
+          "accurate at this depth.");
+    }
+  }
+};
+
+// Helper to get types from treelite
+template <typename tl_threshold_t, typename tl_output_t>
+auto
+get_tree_meta_info_vector(
+    const treelite::ModelImpl<tl_threshold_t, tl_output_t>&)
+{
+  return std::vector<TreeMetaInfo<tl_threshold_t, tl_output_t>>{};
+}
+
+double
+psi(double* e, const double* offset, const double* Base, double q,
+    const double* n, int d)
+{
+  double res = 0.;
+  for (int i = 0; i < d; i++) {
+    res += e[i] * offset[i] / (Base[i] + q) * n[i];
+  }
+  return res / d;
+}
+
+void
+times(const double* input, double* output, double scalar, int size)
+{
+  for (int i = 0; i < size; i++) {
+    output[i] = input[i] * scalar;
+  }
+};
+
+void
+times_broadcast(const double* input, double* output, int size)
+{
+  for (int i = 0; i < size; i++) {
+    output[i] *= input[i];
+  }
+};
+
+void
+addition(double* input, double* output, int size)
+{
+  for (int i = 0; i < size; i++) {
+    output[i] += input[i];
+  }
+};
+
+void
+write(double* from, double* to, int size)
+{
+  for (int i = 0; i < size; i++) {
+    to[i] = from[i];
+  }
+};
+
+
+template <typename ThresholdType>
+bool
+decision_non_categorical(
+    float fvalue, ThresholdType threshold, treelite::Operator op)
+{
+  // XGBoost
```

Comment:
I'm assuming we don't condense this all into one switch statement for perf? If so, could you make a comment to that effect?

Also, if perf is relevant here, any reason not to use a single return statement?

Reply:
This was taken verbatim from treelite. Yes it should be one statement.